updating vcs setup instructions (#917)
hashicorp/terraform-docs-common: website/docs/cloud-docs/vcs/azure-devops-services.mdx

# Set up the Azure DevOps Services VCS provider using OAuth

> ~> **Important:** Starting April 2025, Microsoft will no longer allow new Azure DevOps OAuth app registrations—see the [official announcement](https://devblogs.microsoft.com/devops/no-new-azure-devops-oauth-apps-beginning-february-2025/) for details. HCP terraform now supports access to Azure DevOps via [Microsoft Entra OAuth Application](https://www.microsoft.com/en-us/security/business/identity-access/microsoft-entra-id). In additional, the use of [personal access tokens](/terraform/cloud-docs/vcs/azure-devops-services-pat) is also supported.

Author: @paulanunda what do you think?

> ~> **Important:** Starting April 2025, Microsoft will no longer allow new Azure DevOps OAuth app registrations—see the [official announcement](https://devblogs.microsoft.com/devops/no-new-azure-devops-oauth-apps-beginning-february-2025/) for details. HCP terraform now supports access to Azure DevOps via [Microsoft Entra OAuth Application](https://www.microsoft.com/en-us/security/business/identity-access/microsoft-entra-id). In additional, the use of [personal access tokens](/terraform/cloud-docs/vcs/azure-devops-services-pat) is also supported.

Reviewer: agreed; i'll commit the suggestion

Leave this page open in a browser tab. You will copy values from this page into Azure DevOps in the next step, and in later steps you will continue configuring HCP Terraform.

> ## Step 2: From your Microsoft Entra admin cente, Create a New Microsoft Entra Application

Reviewer 2: Typo: `center`

> ## Step 2: From your Microsoft Entra admin cente, Create a New Microsoft Entra Application

Reviewer 3: ```suggestion
## Step 2: From your Microsoft Entra admin center, Create a New Microsoft Entra Application
```

1. In a new browser tab, login to your [Microsoft Entra admin center](https://entra.microsoft.com), and open the page to register a new MS Entra App by navigating to this [link](https://entra.microsoft.com/#view/Microsoft_AAD_RegisteredApps/CreateApplicationBlade/quickStartType~/null/isMSAApp~/false). If you have access to multiple Entra tenants, switch to the tenant in which you want to register the application.

1. A form appears asking for your company and application information.

> 1. Fill out the fields and checkboxes with the corresponding values currently displayed in your HCP Terraform browser tab. HCP Terraform lists the values in the order they appear and includes controls for copying values to your clipboard. Here is an example:

Reviewer 3: Do we have any guidance on if/when they should change the default selection under "Supported account types", or can they safely leave the default?

> 1. Fill out the fields and checkboxes with the corresponding values currently displayed in your HCP Terraform browser tab. HCP Terraform lists the values in the order they appear and includes controls for copying values to your clipboard. Here is an example:

Author: @lilincmu what do you think about this one, can it be left as default?

   | Field name                 | Value                                                                         |
   | -------------------------- | ----------------------------------------------------------------------------- |
   | Application Name           | HCP Terraform (`<YOUR ORGANIZATION NAME>`)                                    |
   | Redirect URI               | `Web` followed by `https://app.terraform.io/<YOUR CALLBACK URL>`               |

1. Click the register button to create the app.

> 1. Once the app is registered, navigate to the **API permissions** tab, and add **vso.code** and **vso.code_status** permissions.

Reviewer 3: I had to go into the "Azure DevOps" section after clicking "Add a permission" is this still accurate?

> 1. Once the app is registered, navigate to the **API permissions** tab, and add **vso.code** and **vso.code_status** permissions.

Author: Yes, it is!

   ![Azure DevOps Services Screenshot: Required permissions when creating a new application in your Azure DevOps Services Profile](/img/docs/azure-devops-services-entra-application-permissions.png)

1. Next navigate to **Certificates and secrets** tab to generate a client secret.

> 1. Navigate back to the **Overview** tab and leave this page open in a browser tab. In the next step, you will copy and paste the unique **Application (client) ID**, **Directory (tenant) ID** and **Client Secret** from this page.

Reviewer 3: ```suggestion
1. Click **Overview** and leave this page open in a browser tab. In the next step, you will copy and paste the unique **Application (client) ID**, **Directory (tenant) ID** and **Client Secret** from this page.
```

1. In HCP Terraform's **Add VCS Provider** page, paste the text of the **SSH private key** you just created, and click the **Add SSH Key** button.

## Step 5: Connect Azure DevOps Organization to the Microsoft Entra Application.

Ensure that the Azure DevOps organization you intend to link with HCP Terraform is connected to the correct Microsoft Entra app created in Step 2 by following these [instructions](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/connect-organization-to-azure-ad?view=azure-devops).

## Finished

At this point, Azure DevOps Services access for HCP Terraform is fully configured, and you can create Terraform workspaces based on your organization's repositories.

> ### If You Need to migrate existing workspaces to connect to ADO Organization via the new MS Entra App

Reviewer 3: ```suggestion
### Migrate existing workspaces to connect to ADO Organization with the new MS Entra App
```

> Go to the HCP Terraform workspaces associated with the Azure DevOps organization via legacy OAuth application, and update the VCS settings to use the new VCS provider created in Step 3.

Reviewer 3: ```suggestion
If you need to migrate existing workspaces to connect to ADO organization, go to the HCP Terraform workspaces associated with the Azure DevOps organization via legacy OAuth application, and update the VCS settings to use the new VCS provider created in Step 3.
```
